### PR TITLE
NeurIPS2018-AI-for-Prosthetics-Challenge training code

### PR DESCRIPTION
**Pull request by zenghsh3**
- NeurIPS2018-AI-for-Prosthetics-Challenge training code
- add replay_memory to parl.utils

### CODE REVIEW

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md [hunk 1]
```diff
@@ -1,36 +1,92 @@
...
-This folder will contains the code used to train the winning models for the [NeurIPS 2018: AI for Prosthetics Challenge](https://www.crowdai.org/challenges/neurips-2018-ai-for-prosthetics-challenge) along with the resulting models. (Codes of training part is organizing, but the resulting models is available now.)
-
-### Dependencies
+This folder contains the code used to train the winning models for the [NeurIPS 2018: AI for Prosthetics Challenge](https://www.crowdai.org/challenges/neurips-2018-ai-for-prosthetics-challenge) along with the resulting models.
+## Dependencies
 - python3.6
-- [paddlepaddle>=1.2.0](https://github.com/PaddlePaddle/Paddle)
+- [paddlepaddle>=1.2.1](https://github.com/PaddlePaddle/Paddle)
 - [osim-rl](https://github.com/stanfordnmbl/osim-rl)
+- grpcio==1.12.1
```
Comment (TomorrowIsAnOtherDay) [inline examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: recommend adding a convenient link for users: https://grpc.io/docs/quickstart/python.html

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md [hunk 2]
```diff
@@ -1,36 +1,92 @@
...
+- tensorflow (To use tensorboard)
+
+## Result
 
-### Start Testing best models
+| Avg reward of all episodes | Avg reward of complete episodes | Falldown % | Evaluate episodes |
+|----------------------------|---------------------------------|------------|-------------------|
+| 9968.5404                  | 9980.3952                       | 0.0026     | 5000              |
```
Comment (TomorrowIsAnOtherDay) [inline examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: More details can be described here: 500 CPUs * 10 episodes.

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md [hunk 3]
```diff
@@ -1,36 +1,92 @@
...
+#### 2. target speed 3.0 m/s
+
+```bash
+# server
+python .simulator_server.py --port [PORT] --ensemble_num 1 --restore_model_path [RunFastest model] --warm_start_batchs 1000
 
-# Set the random seed 
-python test.py --seed 1024
+# client (Suggest: 200+ clients)
+python simulator_client.py --port [PORT] --ip [IP] --reward_type FixedTargetSpeed --target_v 3.0 --act_penalty_lowerbound 1.5
```
Comment (TomorrowIsAnOtherDay) [inline examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: Though we mentioned that aligning all arguments in one line seems more compact, I found too much words in one line is strange, too. A scroll bar is appeared in the bottom of the line for two many words.
We may split this line into several lines.

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md [hunk 4]
```diff
@@ -1,36 +1,92 @@
...
+
+#### 3. target speed 2.0 m/s
+
+```bash
+# server
+python .simulator_server.py --port [PORT] --ensemble_num 1 --restore_model_path [FixedTargetSpeed 3.0m/s model] --warm_start_batchs 1000
 
-# Set the episode number to run
-python test.py --episode_num 2
+# client (Suggest: 200+ clients)
```
Comment (TomorrowIsAnOtherDay) [inline examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md:64]: This suggestion is great, but could you explain why 200+ clients is a sensible choice ?

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md [hunk 5]
```diff
@@ -1,36 +1,92 @@
...
+# server
+python .simulator_server.py --port [PORT] --ensemble_num 1 --restore_model_path [FixedTargetSpeed 2.0m/s model] --warm_start_batchs 1000 
+
+# client (Suggest: 200+ clients)
+python simulator_client.py --port [PORT] --ip [IP] --reward_type FixedTargetSpeed --target_v 1.25 --act_penalty_lowerbound 0.6
+```
+
+### Stage II: Round2
+
+> You can install resulting 1.25m/s model in Stage I from [Baidu Pan](https://pan.baidu.com/s/1PVDgIe3NuLB-4qI5iSxtKA) or [Google Drive](https://drive.google.com/open?id=1jWzs3wvq7_ierIwGZXc-M92bv1X5eqs7)
```
Comment (TomorrowIsAnOtherDay) [inline examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: install -> download ?

File: parl/utils/replay_memory.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        self.obs = np.zeros((max_size, obs_dim), dtype='float32')
+        self.action = np.zeros((max_size, act_dim), dtype='float32')
+        self.reward = np.zeros((max_size, ), dtype='float32')
+        self.terminal = np.zeros((max_size, ), dtype='bool')
+        self.next_obs = np.zeros((max_size, obs_dim), dtype='float32')
+
+        self._curr_size = 0
+        self._curr_pos = 0
+
+    def sample_batch(self, batch_size):
```
Comment (TomorrowIsAnOtherDay) [inline parl/utils/replay_memory.py:36]: I remember that we have a offset on sampling index, to avoid sampling some dirty data caused by multi-thread processing. In current code, you may sample the data that are writing.

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md (hunk 1 above)
Comment (zenghsh3) [reply examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: done

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md (hunk 2 above)
Comment (zenghsh3) [reply examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: done

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md (hunk 3 above)
Comment (zenghsh3) [reply examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: done

File: examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md (hunk 5 above)
Comment (zenghsh3) [reply examples/NeurIPS2018-AI-for-Prosthetics-Challenge/README.md]: done

File: parl/utils/replay_memory.py (hunk 6 above)
Comment (zenghsh3) [reply parl/utils/replay_memory.py:36]: done

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
Look good to me :)
